### PR TITLE
Implement blocking servlet/tomcat/jetty

### PR DESCRIPTION
**Pull request by cataphract**
Implements blocking behavior for servlet/tomcat/jetty. Blocking means a fixed error response is send and request handling is skipped.

### CODE REVIEW

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  private static final Logger log = LoggerFactory.getLogger(TomcatBlockingHelper.class);
+  private static final int STATUS_CODE = 403;
+  private static final String RESPONSE_TEXT = "Access denied (request blocked)";
+  private static final byte[] RESPONSE_BODY = RESPONSE_TEXT.getBytes(StandardCharsets.US_ASCII);
+  private static final MethodHandle GET_OUTPUT_STREAM;
+
+  static {
+    MethodHandle mh = null;
+    try {
+      Method getOutputStream = Response.class.getMethod("getOutputStream");
```
Comment (dougqh) [inline dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java:22]: Why do the lookup via reflection to then use a method handle?
Why not just use MethodHandles.lookup directly?

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  public static void commitBlockingResponse(Response resp) {
+    if (!start(resp, STATUS_CODE)) {
+      return;
+    }
+
+    resp.setHeader("Content-length", Integer.toString(RESPONSE_BODY.length));
+    resp.setHeader("Content-type", "text/plain");
+    try {
+      OutputStream os = (OutputStream) GET_OUTPUT_STREAM.invoke(resp);
```
Comment (dougqh) [inline dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java:38]: Why do we need reflection to call Response.getOutputStream?

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java (hunk 1 above)
Comment (cataphract) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java:22]: No reason at all, just inartful on my part.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java (hunk 2 above)
Comment (cataphract) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java:38]: The return value changes between `OutputStream` and `ServletOutputStream (extends OutputStream)`, so the method reference changes.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java (hunk 1 above)
Comment (cataphract) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatBlockingHelper.java:22]: Ah, actually the reason is the same I gave in the other comment. When using `lookup` I need to specify the return type, whereas with reflection I don't.

**Comment by cataphract** [comment]
Can only be merged after changes in the WAF/default AppSec configuration. Right now, the system tests cannot pass because the WAF returns a blocking instruction and the system tests expect no blocking.

**Comment by ValentinZakharov** [comment]
I can see usage of `span.markForBlocking()` in case of:
- reading/parsing URL
- reading http headers
- reading socket address

but I can't see such blocking ability for:
- http response (we have few rules that check `server.response.status`)
- parsing request body 

Was it skipped intentionally?

File: dd-java-agent/instrumentation/jetty-common/src/main/java/datadog/trace/instrumentation/jetty9/HandleVisitor.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+ * under a condition.
+ */
+public class HandleVisitor extends MethodVisitor {
+  private static final Logger log = LoggerFactory.getLogger(HandleVisitor.class);
+
+  private boolean lookForStore;
+  private int agentSpanVar = -1;
+  private boolean success;
+  private final String methodName;
+  private boolean foundFirstDispatch;
```
Comment (ValentinZakharov) [inline dd-java-agent/instrumentation/jetty-common/src/main/java/datadog/trace/instrumentation/jetty9/HandleVisitor.java:27]: It seems `foundFirstDispatch` is unused

**Comment by ValentinZakharov** [comment]
I can see usage of `IGSpanInfo.isToBeBlocked()` as a blocking indicator for Servlet/Tomcat/Undertow, but not for Jetty. I don't really understand how is Jetty aware about blocking?

File: dd-java-agent/instrumentation/servlet-common/src/main/java/datadog/trace/instrumentation/servlet/ServletBlockingHelper.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import java.nio.charset.StandardCharsets;
+import javax.servlet.http.HttpServletResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ServletBlockingHelper {
+  private static final Logger log = LoggerFactory.getLogger(ServletBlockingHelper.class);
+  private static final int STATUS_CODE = 403;
+  private static final String RESPONSE_TEXT = "Access denied (request blocked)";
+  private static final byte[] RESPONSE_BODY = RESPONSE_TEXT.getBytes(StandardCharsets.US_ASCII);
```
Comment (ValentinZakharov) [inline dd-java-agent/instrumentation/servlet-common/src/main/java/datadog/trace/instrumentation/servlet/ServletBlockingHelper.java:14]: Blocking info (body, status etc.) are currently duplicated for each blocking instrumentation. Did you consider any general approach to have blocking data in same place?
May be some decorator with defined blocking logic can fit our requirements?

**Review by ValentinZakharov — APPROVED** [review]

**Comment by cataphract** [comment]
> I can see usage of `span.markForBlocking()` in case of:
> 
> * reading/parsing URL
> * reading http headers
> * reading socket address
> 
> but I can't see such blocking ability for:
> 
> * http response (we have few rules that check `server.response.status`)
> * parsing request body
> 
> Was it skipped intentionally?

Yes, it was skipped intentionally. IIRC, the plan is to first only block on ip addresses, and then only on users.

**Comment by cataphract** [comment]
> I can see usage of `IGSpanInfo.isToBeBlocked()` as a blocking indicator for Servlet/Tomcat/Undertow, but not for Jetty. I don't really understand how is Jetty aware about blocking?

It is used for Jetty. See https://github.com/DataDog/dd-trace-java/pull/3701/files#diff-8d4f557f90cc211125aed1eaf1c1232fec4a1b9be60e96959e77098f08add2a9R76-R81

File: dd-java-agent/agent-bootstrap/src/main/resources/datadog/trace/bootstrap/blocking/template.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{"errors": [{"title": "Access denied", "detail": "Your request has been blocked and was not executed."}]}
```
Comment (nizox) [inline dd-java-agent/agent-bootstrap/src/main/resources/datadog/trace/bootstrap/blocking/template.json:1]:
```suggestion
{"errors": [{"title": "You've been blocked", "detail": "Sorry, you cannot access this page. Please contact the customer service team. Security provided by Datadog."}]}
```

File: dd-java-agent/agent-bootstrap/src/main/resources/datadog/trace/bootstrap/blocking/template.html [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    footer p {
+      font-size: 16px
+    }
+  </style>
+</head>
+
+<body>
+<main>
+  <p>Sorry, you cannot access this page. Please contact our custom support team.</p>
```
Comment (nizox) [inline dd-java-agent/agent-bootstrap/src/main/resources/datadog/trace/bootstrap/blocking/template.html:90]:
```suggestion
  <p>Sorry, you cannot access this page. Please contact the customer service team.</p>
```

**Review by smola — APPROVED** [review]
We discussed possible ways to simplify this instrumentation using our new call site instrumentation framework. However, it will not be possible. We'll need:

* Supporting inserting a call site advice only in specific classes (we expect to do this at some point).
* Adding some further facilities to call private methods without reflection.

So, it's not happening (yet).

**Pull request merged**